File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.is_balanced = True

        def dfs(node):
            if node is None:
                return 0
            
            left = dfs(node.left)
            right = dfs(node.right)

            if (abs(left - right) > 1):
                self.is_balanced = False
            
            return 1 + max(left, right)
        
        dfs(root)
        return self.is_balanced

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == nullptr) {
            return nullptr;
        }

        int low = min(p->val, q->val);
        int high = max(p->val, q->val);
        TreeNode *node = root;

        while (node != nullptr) {
            int nVal = node->val;

            if (nVal < low) {
                node = node->right;
            }
            else if (nVal > high) {
                node = node->left;
            }
            else {
                return node;
            }
        }
        return nullptr;
    }
};
